File: framework/languages/dotnet/src/Zlink.Framework/Contracts/Streams/IZLinkSession.cs
namespace Zlink.Framework.Contracts.Streams;

public interface IZLinkSession
{
    IZLinkSessionContext Context { get; }

    ValueTask OnConnectedAsync(
        CancellationToken cancellationToken);

    ValueTask OnDisconnectedAsync(
        CancellationToken cancellationToken);

    ValueTask OnErrorAsync(
        ZLinkStreamError error,
        CancellationToken cancellationToken);

    /// <summary>
    ///     Handles a framework-owned inbound stream payload.
    /// </summary>
    /// <remarks>
    ///     The payload is a framework <see cref="ZLinkMessage" />. Session code may
    ///     decode it or pass it to framework APIs such as
    ///     <see cref="IZLinkSessionActor.RelayAsync" />.
    /// </remarks>
    ValueTask OnDispatchAsync(
        ZLinkSessionDispatchContext dispatch,
        ZLinkMessage payload,
        CancellationToken cancellationToken)
    {
        _ = dispatch;
        _ = payload;
        _ = cancellationToken;
        return ValueTask.CompletedTask;
    }
}

public interface IZLinkSessionClient
{
    IZLinkSessionSendCall Send<TMessage>(TMessage message);

    IZLinkSessionReplyCall Reply<TMessage>(TMessage message);
}

public interface IZLinkSessionActors
{
    IReadOnlyCollection<IZLinkSessionActor> Bound { get; }

    ValueTask<IZLinkSessionActor> BindAsync(
        IZLinkActor actor,
        CancellationToken cancellationToken = default);

    ValueTask<IZLinkSessionActor> BindAsync(
        ActorRef actor,
        CancellationToken cancellationToken = default);

    IZLinkSessionActor? Find(string actorId);
}

public interface IZLinkSessionContext
{
    string SessionId { get; }

    RoutingId? RoutingId { get; }

    string? LocalAddr { get; }

    string? RemoteAddr { get; }

    IZLinkSessionClient Client { get; }

    IZLinkSessionActors Actors { get; }

    ValueTask CloseAsync();
}

public interface IZLinkSessionSendCall
{
    IZLinkSessionSendCall Metadata(string key, string value);

    IZLinkSessionSendCall PacketName(string messageName);

    IZLinkSessionSendCall Compress();

    ValueTask Async();
}

public interface IZLinkSessionReplyCall
{
    IZLinkSessionReplyCall Metadata(string key, string value);

    IZLinkSessionReplyCall Compress();

    ValueTask Async();
}

File: framework/languages/node/packages/framework/src/contracts/Streams/IZLinkSession.ts
import type { RoutingId, ZLinkMessage } from '../Common';
import type { ZLinkSessionActors } from './IZLinkSessionActor';
import type { ZLinkStream } from './IZLinkStream';
import type { ZLinkStreamError } from './ZLinkStreamError';

export interface ZLinkSession {
  readonly context: ZLinkSessionContext;
  onConnected?(context: ZLinkSessionContext): Promise<void>;
  onDisconnected?(context: ZLinkSessionContext): Promise<void>;
  onError?(context: ZLinkSessionContext, error: ZLinkStreamError): Promise<void>;
  onDispatch?(dispatch: ZLinkSessionDispatchContext, payload: ZLinkMessage, signal?: AbortSignal): Promise<void>;
}

export interface ZLinkSessionDispatchContext {
  readonly packetName: string;
  readonly metadata: ReadonlyMap<string, string>;
  readonly canReply: boolean;
}

export interface ZLinkSessionFactory<TSession extends ZLinkSession = ZLinkSession> {
  create(context: ZLinkSessionContext): TSession | Promise<TSession>;
}

export interface ZLinkSessionContext {
  readonly sessionId: string;
  readonly routingId?: RoutingId;
  readonly localAddr?: string;
  readonly remoteAddr?: string;
  readonly stream: ZLinkStream;
  readonly client: ZLinkSessionClient;
  readonly actors: ZLinkSessionActors;
  close(signal?: AbortSignal): Promise<void>;
}

export interface ZLinkSessionClient {
  send(message: unknown): ZLinkSessionSendCall;
  reply(message: unknown): ZLinkSessionReplyCall;
}

export interface ZLinkSessionSendCall {
  metadata(key: string, value: string): this;
  packetName(packetName: string): this;
  compress(enabled?: boolean): this;
  submit(signal?: AbortSignal): void;
}

export interface ZLinkSessionReplyCall {
  metadata(key: string, value: string): this;
  compress(enabled?: boolean): this;
  submit(signal?: AbortSignal): void;
}
